### DISCUSSION TITLE
Valetudo 2022.10.0

### DISCUSSION THREAD

**Discussion by Hypfer** [Hypfer/Valetudo#1615] [Release Discussions]
<div align="center">
    <img src="https://raw.githubusercontent.com/Hypfer/Valetudo/master/assets/logo/valetudo_logo_with_name.svg" width="800" alt="valetudo">
    <p align="center"><h2>2022.10.0</h2></p>
</div>

New Robot, new Firmware, Various UX improvements and MQTT changes

## Dreame W10 support

[<img src="https://user-images.githubusercontent.com/974410/196250070-c6f50089-e239-41ab-a065-73602280ac81.png" width=700>](https://user-images.githubusercontent.com/974410/196250070-c6f50089-e239-41ab-a065-73602280ac81.png)

Starting with this release, the Dreame W10 (not-pro!) is supported by Valetudo.
Because I got pretty annoyed by the lack of proper mopping robots supported by Valetudo, I've finally just bought one myself to check if it is rootable and suitable for Valetudo.

Turns out:
It is. Looks like it always was.
We were just misinformed because we never had bought one ourselves due to its pricing and people on the internet told us incorrect information about it. The lesson here is that we absolutely have to buy _all_ robots ourselves or else we go nowhere.

Anyway, rooting is possible with the catch that you will need a new custom PCB to do so.
I do have ~250 of them lying here waiting to be soldered and used, however the logistical issue of me getting those to you yet remains to be solved.

## Dreame D9 Pro firmware backport

Because I got pretty annoyed by the lack of firmware updates for the D9 Pro, I've finally just bought one myself to check if I can do anything about it.

It seemed pretty wrong that the older D9 has a recent 2022 firmware while the newer D9 Pro is stuck with a firmware from early 2021.
Thus, I wondered how different those robots really are and if it would be possible to port the D9 firmware over to it.

Turns out:
They're pretty similar but not entirely.

And so with a few changes, you can now upgrade your D9 Pro to a much more recent firmware using the Dustbuilder.
Gone should be the days of the robot bumping into everything and its map slowly deteriorating over time.

You're welcome :)

![image](https://user-images.githubusercontent.com/974410/196250215-3c4a9743-210f-46dc-a454-abbc80430083.png)

**Reminder:**
**Do not try to install a regular D9 firmware on your D9 Pro**. It will **brick your robot**.
You **absolutely need** to select the **customized D9 Pro version** in the dustbuilder.

## MQTT

This release features a few MQTT improvements.

### Optional exposed capabilities

Valetudo's MQTT client has gotten a bit more flexible with this release.

To avoid additional load and traffic for features that aren't required in 90% of setups, not every capability is exposed to MQTT.
This unfortunately meant that 10% of people had to also use the REST API to solve their automation needs.

![image](https://user-images.githubusercontent.com/974410/196250308-72e7757b-38e3-47ed-9aa4-637fa32577a2.png)

With optional exposed capabilities, it is possible to cater to those needs without polluting the MQTT broker of people that don't require the functionality. For now, the only option to select is the SpeakerVolumeControlCapability. Others _may_ follow if someone makes a solid case for why it should be controllable via MQTT.

![image](https://user-images.githubusercontent.com/974410/196250338-2b00bd29-d50c-499f-ad55-e8288c8211b3.png)

Please only enable optional exposed capabilities that you actually plan on using as each additional one adds system load due to periodic polling.

### Improved MQTT Errors

The error mapping improvements that have been implemented in the last release are now also being published to MQTT.

[<img src="https://user-images.githubusercontent.com/974410/196250399-1afb0ded-17c8-4fb5-9a8d-2664631a4851.png" width=700>](https://user-images.githubusercontent.com/974410/196250399-1afb0ded-17c8-4fb5-9a8d-2664631a4851.png)

It should make building reliable automations based on the error state much easier

Please note that this could be a **breaking change** if you've hardcoded the MQTT error topic somewhere.
Users of Home Assistant should not notice anything changing as the autoconfiguration payload change should take care of everything.

If you've created automations based on the value of the error topic, this will be a **breaking change** for those as well.

### Better map integration instructions

The process of getting the Valetudo Map into Home Assistant always has been challenging for newcomers as it is not straightforward.
Therefore, the image displayed in Home Assistant was changed and a new docs microsite was created.

[<img src="https://user-images.githubusercontent.com/974410/196250481-9a399468-912b-4f37-9b43-d604e731624d.png" width=400>](https://user-images.githubusercontent.com/974410/196250481-9a399468-912b-4f37-9b43-d604e731624d.png)

If you navigate to hass.valetudo.cloud, you'll find a step-by-step walkthrough to get the map integrated.
Furthermore, there is also an explanation why things are that complicated.

## UX Improvements

[<img src="https://user-images.githubusercontent.com/974410/196250561-5b763120-0b5d-46e4-b248-0319f18dd4c6.png" width=400>](https://user-images.githubusercontent.com/974410/196250561-5b763120-0b5d-46e4-b248-0319f18dd4c6.png)

The Mop Attachment Reminder event now gets dismissed automatically when you detach the Mop Attachment from the robot.

[<img src="https://user-images.githubusercontent.com/974410/196250620-359902c8-6419-4339-90d8-a8479904f737.png" width=700>](https://user-images.githubusercontent.com/974410/196250620-359902c8-6419-4339-90d8-a8479904f737.png)

If the user attempts to start a full cleanup after selecting something in the map view, they will now see a warning dialog.

[<img src="https://user-images.githubusercontent.com/974410/196250696-3cebd17a-c9f9-4ecd-9eb0-8b1ba26121e1.png" width=400>](https://user-images.githubusercontent.com/974410/196250696-3cebd17a-c9f9-4ecd-9eb0-8b1ba26121e1.png)

The Wi-Fi configuration settings are now hidden and replaced by an info box if the robot does not support reconfiguration while provisioned.

## Operation Mode Control

This release features support for changing the operation mode of the robot.

![image](https://user-images.githubusercontent.com/974410/196250744-6b4be140-47bc-457f-8f0a-62b885c0bfa7.png)

This enables you to e.g., tell your new W10 that it should only mop.

## More error code mappings

After staring at Ghidra for a while, I've managed to discover the meaning of a few more hardware error codes of dreame robots.
Furthermore, roborock hardware error code mappings were updated as well.

## The usual

As you've seen at the beginning of these release notes, we do have to buy all robots ourselves if we want stuff to happen.

Thus, if you want to see Valetudo on more robots and/or like this release, you might want to consider donating if you haven't done so already:

[https://github.com/sponsors/Hypfer](https://github.com/sponsors/Hypfer)

[https://builder.dontvacuum.me/donations.txt](https://builder.dontvacuum.me/donations.txt)

### Breaking Changes

- **mqtt**: Provide better error information [`da6994a`](http://github.com/Hypfer/Valetudo/commit/da6994a34de68654a35daf9819936fa20cc12539)

### Features

- **vendor.dreame**: Newer 1C firmwares do seem to support no mop zones [`a06a86d`](http://github.com/Hypfer/Valetudo/commit/a06a86d3e542429232714dc932702bf2a0703c74)
- **vendor.dreame**: Operation mode control [`aefa436`](http://github.com/Hypfer/Valetudo/commit/aefa436dafb590ba13c8d1fd36751d6918d5a93f)
- **vendor.dreame**: D9 Pro+ special edition support [`ffbcb73`](http://github.com/Hypfer/Valetudo/commit/ffbcb736e45ad156be67b8565d1f6e1a9e1c93e0)
- **vendor.dreame**: Mop Stuff [`0081cba`](http://github.com/Hypfer/Valetudo/commit/0081cbaad4be58e3a0d220408f211f013e776cc0)
- **ui**: Allow configuration of optional exposable capabilities for mqtt [`3f31186`](http://github.com/Hypfer/Valetudo/commit/3f31186fd0bf8578252b47dab586525a0b6b1c96)
- **mqtt**: Introduce optional exposed capabilities starting with the SpeakerVolumeControlCapabilityMqttHandle [`685227b`](http://github.com/Hypfer/Valetudo/commit/685227b226505ea669a2a53ef2f70386dee37580)
- **vendor.dreame**: Initial support for the W10 [`6da6556`](http://github.com/Hypfer/Valetudo/commit/6da6556644b145896add18e221c19f5c2b79f082)
- **ui**: Warn the user when attempting to start a full cleanup with a MapAction pending [`bc6a4ce`](http://github.com/Hypfer/Valetudo/commit/bc6a4ce18c542f3f38feeefa273a7a375e502bed)
- **ui**: Add current browser timezone hint to timer timePicker [`892cc09`](http://github.com/Hypfer/Valetudo/commit/892cc096919ce3aca4382497416ce36c62a03749)
- **vendor.dreame**: Automatically dismiss the MopAttachmentReminderValetudoEvent on Mop detachment [`19f96a2`](http://github.com/Hypfer/Valetudo/commit/19f96a2ee09cc690af375995117f997037c77189)
- **mqtt**: Update wrapper image to point to docs microsite [`67f600a`](http://github.com/Hypfer/Valetudo/commit/67f600a4e0f1c3aea67c8a6c968acec2d911ee46)
- **vendor.dreame**: Map some internal error codes [`c9bc1b8`](http://github.com/Hypfer/Valetudo/commit/c9bc1b85e3afdd9d79263ee7fef7c084f3beedf3)
- **vendor.roborock**: Add hardware fault error code mappings [`19ce6aa`](http://github.com/Hypfer/Valetudo/commit/19ce6aad052dbccbb04e9ff539bef0dd5af75937)

### Fixes

- **vendor.dreame**: Fix generational differences in operation mode mapping [`8b6643c`](http://github.com/Hypfer/Valetudo/commit/8b6643c978b979aaddfa5806ac5cfba130f441e8)
- **vendor.dreame**: 1C error codes might occasionally be numbers [`93fdfb9`](http://github.com/Hypfer/Valetudo/commit/93fdfb927f0fbbba1b5eabcc15a3a858d97281d3)
- Hide non-functional Wi-Fi configuration editor if not supported [`fbcfd47`](http://github.com/Hypfer/Valetudo/commit/fbcfd47051af37d17d8de7640adb318620dc8790)
- **vendor.dreame**: Handle carpet information in maps [`d1ad1af`](http://github.com/Hypfer/Valetudo/commit/d1ad1aff916ddce260dbd65f231149c3ae8687fc)
- **ui**: Strip even more characters that home assistant doesn't like from all mqtt topic building blocks [`5877381`](http://github.com/Hypfer/Valetudo/commit/587738132a2d7a5bb4c8b4ca0b60d001f50a550a)
- **vendor.dreame**: Voicepack hash must be lowercase [`0deff82`](http://github.com/Hypfer/Valetudo/commit/0deff822f29962eab2afea92b7e881c31b386397)
- **ui**: Weekday order should be ISO 8601 compliant [`3c884f5`](http://github.com/Hypfer/Valetudo/commit/3c884f58c424f7e173815bc2a1f89bd7bdbdd95f)

### Refactoring

- Misc cleanup [`9fa98df`](http://github.com/Hypfer/Valetudo/commit/9fa98dfdacf478534d944f49471d39c6c0eda95f)

### Chores

- **release**: 2022.10.0 [`dd49ecc`](http://github.com/Hypfer/Valetudo/commit/dd49ecc097fefcf650e5774ff36671e17911e620)
- **mqtt**: Minor cleanup [`d27b189`](http://github.com/Hypfer/Valetudo/commit/d27b18932ae2f3d35cadd21614fd4336c6ac620f)
- Bump github actions to setup-node@v3 [`4dbb745`](http://github.com/Hypfer/Valetudo/commit/4dbb745e00538f2e3be8d2275413c68230c02701)
- bump upx dependency [`38fb6e4`](http://github.com/Hypfer/Valetudo/commit/38fb6e43eff6b158b1d18e09259910d2341758b0)
- Revert axios bump due to broken builds [`c3de7e5`](http://github.com/Hypfer/Valetudo/commit/c3de7e505b2a9ea3e0323e4131197a34ed0a24fc)
- Bump dependencies [`cefd3a7`](http://github.com/Hypfer/Valetudo/commit/cefd3a77196cc1b151551e3eef910ffe5d99c377)
- **webserver**: Minor data validation cleanup [`263774b`](http://github.com/Hypfer/Valetudo/commit/263774b845d627b6c6c589614a79198a79592f11)
- **vendor.dreame**: Ignore irrelevant POST_CHARGE_CONTINUE update on initial root [`0fdec0c`](http://github.com/Hypfer/Valetudo/commit/0fdec0c7922a1dab2e613e254144a3352ce6c349)
- Docs update [`a96501f`](http://github.com/Hypfer/Valetudo/commit/a96501f972e351ec9f2a7cdab51618bb3c4172bb)
- **mqtt**: Update wrapper image [`cd3b75f`](http://github.com/Hypfer/Valetudo/commit/cd3b75f8550acbba3bfeedda1bc2967adb995881)

<hr /><em>This discussion was created from the release <a href='https://github.com/Hypfer/Valetudo/releases/tag/2022.10.0'>Valetudo 2022.10.0</a>.</em>

**Comment by l3ztum** [comment]
@Hypfer any chance to get the gerber files of the pcb for the dreame w10 or better one of the 250 PCBs? i'm situated in germany.

**Reply by Numior** [reply]
Have the same interest. Of course I would pay for the PCB and shipping costs if necessary. Thanks!

**Reply by pfoff** [reply]
I'd also love to get one of those pcbs. For now:  can anyone tell me where w10's uart connector is located?

**Reply by klonfish** [reply]
@pfoff The debug port is under a small lid at the front of the dust bin compartment. You can see it in Dennis' teardown pictures: https://dontvacuum.me/teardowns/dreame.vacuum.p2027/
If I'm not mistaken, the pins required for UART should be on the top row near the left (side brush) side. Think of the connector as shown in the "Rooting instructions" just tilted towards the back of the robot. As always, please be careful and check if this information fits to your robot (e.g. by checking if the two GND pins are where they are supposed to be).

**Reply by pfoff** [reply]
Thanks a lot, I found it! 
So I will build me a little adapter and goin' to check the GND pins first. 
You were right, I got a connection like that. Now I get this console output. Seems to get more complicated to get it rooted? Any hints where to continue from here on?-(
![dreame_console_2022-12-06_23-26_](https://user-images.githubusercontent.com/25942044/206038291-2cb5a0ca-5c2a-4fc9-93e8-7963609be2de.png)

**Reply by AlexH-HankIT** [reply]
Looks like i'm facing the same issue as you @pfoff . Did you figure out a solution?

**Reply by pfoff** [reply]
Nope, I created a ticket on vaers.
